syntax = "proto3";

package collection_service;

service CollectionService {
    rpc GetCollections (Empty) returns (ListCollectionResponse);
    rpc GetCollectionById (CollectionRequest) returns (CollectionResponse);
    rpc CreateCollection (CreateCollectionRequest) returns (CollectionResponse);
    rpc UpdateCollection (UpdateCollectionRequest) returns (CollectionResponse);
    rpc DeleteCollection (DeleteCollectionRequest) returns (DeleteCollectionResponse);
    rpc AddRecipeToCollection (ModifyRecipeRequest) returns (ModifyRecipeResponse);
    rpc RemoveRecipeFromCollection (ModifyRecipeRequest) returns (ModifyRecipeResponse);
}

message Empty {}

message CollectionRequest {
    int32 id = 1;
}

message CreateCollectionRequest {
    string name = 1;
    string author = 2;
    string description = 3;
    repeated string recipes = 4;
}

message UpdateCollectionRequest {
    int32 id = 1;
    string author = 2;
    string name = 3;
    string description = 4;
    repeated string recipes = 5;
}

message DeleteCollectionRequest {
    int32 id = 1;
    string author = 2;
}

message ModifyRecipeRequest {
    int32 id = 1;
    string recipe_id = 2;
}

message CollectionResponse {
    int32 id = 1;
    string name = 2;
    string author = 3;
    string description = 4;
    repeated string recipes = 5;
}

message ListCollectionResponse {
    repeated CollectionResponse collections = 1;
}

message DeleteCollectionResponse {
    string status = 1;
}

message ModifyRecipeResponse {
    string status = 1;
}